src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -186,54 +223,65 @@ async def _server_call_request_call(Server server,
     return rpc_state
 
 
-async def _server_main_loop(Server server,
-                            _CallbackCompletionQueue cq,
-                            list generic_handlers):
-    cdef object loop = asyncio.get_event_loop()
-    cdef RPCState rpc_state
-
-    while True:
-        rpc_state = await _server_call_request_call(
-            server,
-            cq,
-            loop)
-
-        loop.create_task(_handle_rpc(generic_handlers, rpc_state, loop))
-
-
-async def _server_start(Server server,
-                        _CallbackCompletionQueue cq,
-                        list generic_handlers):
-    server.start()
-    await _server_main_loop(server, cq, generic_handlers)
+cdef _CallbackFailureHandler CQ_SHUTDOWN_FAILURE_HANDLER = _CallbackFailureHandler(
+    'grpc_completion_queue_shutdown',
+    'Unknown',
+    RuntimeError)
 
 
 cdef class _CallbackCompletionQueue:
 
-    def __cinit__(self):
+    def __cinit__(self, object loop):
+        self._loop = loop
+        self._shutdown_completed = loop.create_future()
+        self._wrapper = CallbackWrapper(
+            self._shutdown_completed,
+            CQ_SHUTDOWN_FAILURE_HANDLER)
         self._cq = grpc_completion_queue_create_for_callback(
-            NULL,
+            self._wrapper.c_functor(),
             NULL
         )
 
     cdef grpc_completion_queue* c_ptr(self):
         return self._cq
+    
+    async def shutdown(self):
+        grpc_completion_queue_shutdown(self._cq)
+        await self._shutdown_completed
+        grpc_completion_queue_destroy(self._cq)
+
+
+cdef _CallbackFailureHandler SERVER_SHUTDOWN_FAILURE_HANDLER = _CallbackFailureHandler(
+    'grpc_server_shutdown_and_notify',
+    'Unknown',
+    RuntimeError)
 
 
 cdef class AioServer:
 
-    def __init__(self, thread_pool, generic_handlers, interceptors, options,
-                 maximum_concurrent_rpcs, compression):
+    def __init__(self, loop, thread_pool, generic_handlers, interceptors,
+                 options, maximum_concurrent_rpcs, compression):
+        # NOTE(lidiz) Core objects won't be deallocated automatically.
+        # If AioServer.shutdown is not called, those objects will leak.
         self._server = Server(options)
-        self._cq = _CallbackCompletionQueue()
-        self._status = AIO_SERVER_STATUS_READY
-        self._generic_handlers = []
+        self._cq = _CallbackCompletionQueue(loop)
         grpc_server_register_completion_queue(
             self._server.c_server,
             self._cq.c_ptr(),
             NULL
         )
+
+        self._loop = loop
+        self._status = AIO_SERVER_STATUS_READY
+        self._generic_handlers = []
         self.add_generic_rpc_handlers(generic_handlers)
+        self._serving_task = None
+
+        self._shutdown_lock = asyncio.Lock()
```

Comment:
Almost all of the keyword argument `loop=` in the surface of `asyncio` is marked deprecated in 3.8, and will be removed in 3.10. E.g. https://docs.python.org/3/library/asyncio-task.html

Reply:
Then IMO we should remove `self._loop = loop` and use the `asyncio.get_event_loop()` when it is used for create a future. Otherwise, we are in a kind of a middle-ground which is not consistent.

In any case, we might end up in the situation in the next future where we decide to add the `loop` kw into the `Channel` and `Server`. But this is a discussion for later on.

Reply:
I'm still thinking that for consistency if you are creating futures with `self._loop` you should also call all Asyncio primitives bypassing the `loop=self._loop` otherwise, you are opening the door on using two different loops.

Or basically remove the `self._loop = loop`, so using always the default one.

In any case, as we already said we are gonna have this discussion again in the future, so up to you.

Reply:
I did miss that point, I will remove it after your PR is merged.

Reply:
I was biased about this topic due to calling `asyncio.get_event_loop` seems verbose and slow. In fact, it is not, it only take 28 ns to finish which is much less than average 70+ ns overhead for normal Python functions. (It is still verbose, but not too slow.)

```
Spent 0.28s for 10000000 runs of get_event_loop
28.40 ns
```

However, if we use variable directly, we can avoid even paying the 28 ns. This might sound strange, but I switched back to your proposal which is using the deprecating `loop=` parameter for AsyncIO primitives.